crates/store/re_query2/src/range.rs

```diff
@@ -0,0 +1,317 @@
+            }
+            .into())
+        }
+    }
+}
+
+impl RangeResults {
+    #[doc(hidden)]
+    #[inline]
+    pub fn add(&mut self, component_name: ComponentName, chunks: Vec<Chunk>) {
+        self.components.insert(component_name, chunks);
+    }
+}
+
+// --- Cache implementation ---
+
+/// Caches the results of `Range` queries for a given [`CacheKey`].
+pub struct RangeCache {
+    /// For debugging purposes.
+    pub cache_key: CacheKey,
+
+    /// All the [`Chunk`]s currently cached.
+    ///
+    /// See [`RangeCachedChunk`] for more information.
+    pub chunks: HashMap<ChunkId, RangeCachedChunk>,
+
+    /// Every [`ChunkId`] present in this set has been asynchronously invalidated.
+    ///
+    /// The next time this cache gets queried, it must remove any entry matching any of these IDs.
+    ///
+    /// Invalidation is deferred to query time because it is far more efficient that way: the frame
+    /// time effectively behaves as a natural micro-batching mechanism.
+    pub pending_invalidations: BTreeSet<ChunkId>,
+}
+
+impl RangeCache {
+    #[inline]
+    pub fn new(cache_key: CacheKey) -> Self {
+        Self {
+            cache_key,
+            chunks: HashMap::default(),
+            pending_invalidations: BTreeSet::default(),
+        }
+    }
+
+    /// Returns the time range covered by this [`RangeCache`].
+    ///
+    /// This is extremely slow (`O(n)`), don't use this for anything but debugging.
+    #[inline]
+    pub fn time_range(&self) -> ResolvedTimeRange {
+        self.chunks
+            .values()
+            .filter_map(|cached| {
+                cached
+                    .chunk
+                    .timelines()
+                    .get(&self.cache_key.timeline)
+                    .map(|time_chunk| time_chunk.time_range())
+            })
+            .fold(ResolvedTimeRange::EMPTY, |mut acc, time_range| {
+                acc.set_min(TimeInt::min(acc.min(), time_range.min()));
+                acc.set_max(TimeInt::max(acc.max(), time_range.max()));
+                acc
+            })
+    }
+}
+
+impl std::fmt::Debug for RangeCache {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let Self {
+            cache_key,
+            chunks,
+            pending_invalidations: _,
+        } = self;
+
+        let mut strings: Vec<String> = Vec::new();
+
+        strings.push(format!(
+            "{} ({})",
+            cache_key.timeline.typ().format_range_utc(self.time_range()),
+            re_format::format_bytes(chunks.total_size_bytes() as _),
+        ));
+
+        if strings.is_empty() {
+            return f.write_str("<empty>");
+        }
+
+        f.write_str(&strings.join("\n").replace("\n\n", "\n"))
+    }
+}
+
+pub struct RangeCachedChunk {
+    pub chunk: Chunk,
+
+    /// When a `Chunk` gets cached, it is pre-processed according to the current [`CacheKey`],
+    /// e.g. it is time-sorted on the appropriate timeline.
```

Comment:
Although this is important for scalar plots it doesn't seem like it is necessarily required for all views of all data.

In particular if all rows contain all data (making the range-join trivial) for many visualizers you could actually traverse the data in an arbitrary order.

Reply:
We're caching the entire `Chunk` here, but at query-time we might only be interested in a small range of data within that `Chunk`.

The caller might not care about the data being sorted or not, but the query engine always cares about finding the sub-range as fast as possible.